agent/consul/gateways/controller_gateways.go

```diff
@@ -2,201 +2,563 @@ package gateways
+	for _, entry := range entries {
+		gateway := entry.(*structs.APIGatewayConfigEntry)
+		for _, listener := range gateway.Listeners {
+			for _, certificate := range listener.TLS.Certificates {
+				if certificate.IsSame(&structs.ResourceReference{
+					Kind:           req.Kind,
+					Name:           req.Name,
+					EnterpriseMeta: *req.Meta,
+				}) {
+					requests = append(requests, controller.Request{
+						Kind: structs.APIGateway,
+						Name: gateway.Name,
+						Meta: &gateway.EnterpriseMeta,
+					})
+				}
+			}
 		}
-		return nil
 	}
+	r.controller.Enqueue(requests...)
+	return nil
+}
 
-	r.ensureBoundGateway(metaGateway)
+func (r apiGatewayReconciler) cleanupBoundGateway(ctx context.Context, req controller.Request, store *state.Store) error {
+	logger := r.logger.With("bound-gateway", req.Name, "partition", req.Meta.PartitionOrDefault(), "namespace", req.Meta.NamespaceOrDefault())
+	logger.Debug("cleaning up bound gateway")
+	defer logger.Debug("finished cleaning up bound gateway")
 
-	routes, err := r.retrieveAllRoutesFromStore()
+	routes, err := retrieveAllRoutesFromStore(store)
 	if err != nil {
 		return err
 	}
 
-	boundGateways, routeErrors := BindRoutesToGateways([]*gatewayMeta{metaGateway}, routes...)
-
-	//In this loop there should only be 1 bound gateway returned, but looping over all returned gateways
-	//to make sure nothing gets dropped and handle case where 0 gateways are returned
-	for _, boundGateway := range boundGateways {
-		// now update the gateway state
-		r.logger.Debug("persisting gateway state", "state", boundGateway)
-		if err := r.store.Update(boundGateway); err != nil {
-			msg := "error persisting state"
-			r.logger.Error(msg, "error", err)
-			return errors.Wrap(err, msg)
-		}
-
-		// then update the gateway status
-		r.logger.Debug("persisting gateway status", "gateway", metaGateway.Gateway)
-		if err := r.store.UpdateStatus(metaGateway.Gateway, err); err != nil {
-			return err
-		}
-	}
+	resource := requestToResourceRef(req)
+	resource.Kind = structs.APIGateway
 
-	// and update the route statuses
-	for route, routeError := range routeErrors {
-		configEntry := r.resourceReferenceToBoundRoute(route)
-		r.logger.Error("route binding error:", routeError)
-		if err := r.store.UpdateStatus(configEntry, routeError); err != nil {
+	for _, toUpdate := range RemoveGateway(resource, routes...) {
+		if err := r.updater.Update(toUpdate); err != nil {
 			return err
 		}
 	}
-
 	return nil
 }
 
-func (r *apiGatewayReconciler) retrieveAllRoutesFromStore() ([]structs.BoundRoute, error) {
-	tcpRoutes, err := r.store.GetConfigEntriesByKind(structs.TCPRoute)
+func (r apiGatewayReconciler) reconcileBoundGateway(ctx context.Context, req controller.Request, store *state.Store, bound *structs.BoundAPIGatewayConfigEntry) error {
+	// this reconciler handles orphaned bound gateways at startup, it just checks to make sure there's still an existing gateway, and if not, it deletes the bound gateway
```

Comment:
As this comment mentions, the main use-case for adding this as a reconciliation loop is if we have orphaned resources due to a leader election switch or controller crash after a gateway has been deleted. If a gateway has been deleted but we don't detect it due to a crash/leader switch, then we'd leave around bound gateway resources and never mark the routes as now unbound from a gateway.

Due to the startup characteristics of the controller we do a list + watch operation when subscribing to a given resource type, so when the controller starts back up this will detect any `BoundAPIGatewayConfigEntry` objects that don't have corresponding APIGateways and garbage collect them.